### PR TITLE
Comprehensions and immutable arrays

### PR DESCRIPTION
**Pull request by antalsz**
This PR adds support for comprehensions and immutable arrays.

I wasn't sure who to request for review for this so I requested a mötley crüe of folks; please feel free to pass on review if you haven't been well-chosen!

### CODE REVIEW

File: lib/Ast.ml [hunk 1]
```diff
@@ -1304,6 +1322,18 @@ end = struct
     let dump {ctx; ast= cl} = dump ctx (Cl cl) in
     assert_no_raise ~f:check_cl ~dump xcl
 
+  let check_comprehension
```
Comment (goldfirere) [inline lib/Ast.ml:1329]: This doesn't quite look right to me.

While I'm unfamiliar with this code, this looks like it's a check to see that when you have a `{ctx;ast}` pair, the `ast` is really one step below the `ctx`. But this `check_comprehension` looks much further than one step. And it seems confused between patterns and expressions: In the pattern case, the expression function is `fun _ -> true`. But that looks like it will accept any comprehension, because every comprehension contains a body.

To me, the right way forward is to make `Ast.t` more glorious, adding cases for comprehension iterators, etc., so that this can be done correctly. Maybe it's possible to proceed without that, but I do think you'll need at least separate functions for the pattern case than for the expression case.

(I guess skipping the glorious `Ast.t` is OK. After all, it's allowed for e.g. an `ast` to be in a list in the `ctx`. You could view all of this iterator stuff as just a structure holding expressions and patterns.)

File: lib/Ast.ml [hunk 2]
```diff
@@ -2384,7 +2510,28 @@ end = struct
     | Exp {pexp_desc= Pexp_field (e, _); _}, {pexp_desc= Pexp_construct _; _}
       when e == exp ->
         true
-    | Exp {pexp_desc; _}, _ -> (
+    | Exp ({pexp_desc; _} as ctx_exp), _ -> (
+      (* JST: This is the old fallthrough case, but we need it for the
```
Comment (goldfirere) [inline lib/Ast.ml]: Why the JST prefix? Especially with `jst_fallthrough_case`, this code looks like it's meant to remain internal to Jane Street... but it's not, I don't think.

File: lib/Ast.ml [hunk 3]
```diff
@@ -2454,19 +2601,28 @@ end = struct
       | Pexp_sequence (lhs, rhs) -> exp_in_sequence lhs rhs exp
       | Pexp_apply (_, args)
         when List.exists args ~f:(fun (_, e0) ->
+                 match Extensions.Expression.of_ast e0 with
```
Comment (goldfirere) [inline lib/Ast.ml:2617]: What's going on here? Can you give an example in a comment?

File: test/passing/tests/attributes.ml [hunk 4]
```diff
@@ -82,7 +82,12 @@ if [@test] true then () else () ;;
 
 if [@test] true then () else if [@test] true then () else ()
 
-let _ = ((A [@test]), (() [@test]), ([] [@test]), [||] [@test])
+let _ = ((A [@test]), (() [@test]), ([] [@test]), [||] [@test], [::] [@test])
```
Comment (goldfirere) [inline test/passing/tests/attributes.ml:85]: This looks odd: parens around the empty list, but not empty arrays? That looks like it predates this patch, though.

File: test/passing/tests/break_separators-after.ml.ref [hunk 5]
```diff
@@ -89,6 +89,14 @@ let _ =
       0
   | _ -> 1
 
+let _ =
+  match something with
+  | [: very_very_long_field_name_running_out_of_space;
```
Comment (goldfirere) [inline test/passing/tests/break_separators-after.ml.ref:94]: Elsewhere, the semicolons come at the beginning. Should I be concerned?

File: test/passing/tests/break_separators-before_docked_wrap.ml.ref [hunk 6]
```diff
@@ -98,6 +98,15 @@ let _ =
       0
   | _ -> 1
 
+let _ =
+  match something with
+  | [:
+     very_very_long_field_name_running_out_of_space
+   ; another_very_very_long_field_name_running_out_of_space; _
```
Comment (goldfirere) [inline test/passing/tests/break_separators-before_docked_wrap.ml.ref:105]: This looks odd to me. Should the `; _` be on its own line?

File: test/passing/tests/break_separators-before_docked_wrap.ml.ref [hunk 7]
```diff
@@ -178,13 +187,57 @@ let length =
   |]
   [@foo]
 
+(* this is an immutable array *)
+let length =
+  [:
+     0; 269999999999999999999999999999999999999999999999999; 26
```
Comment (goldfirere) [inline test/passing/tests/break_separators-before_docked_wrap.ml.ref:193]: Also odd. But it looks like it's in keeping with other examples nearby that are unchanged....

File: vendor/diff-parsers-ext-parsewyc.patch [hunk 8]
```diff
@@ -578,6 +680,21 @@
    | LET OPEN override_flag attributes mkrhs(mod_longident) IN class_signature
        { let loc = ($startpos($2), $endpos($5)) in
          let od = Opn.mk ~override:$3 ~loc:(make_loc loc) $5 in
+@@@@
+       { $1 }
+   | poly_pattern
+       { $1 }
+ ;
+ %inline poly_pattern:
+-    mkpat(
```
Comment (goldfirere) [inline vendor/diff-parsers-ext-parsewyc.patch:689]: This looks like a spacing-only change. Possible to remove?

File: vendor/diff-parsers-ext-parsewyc.patch [hunk 9]
```diff
@@ -178,6 +178,22 @@
  
  let make_loc (startpos, endpos) = {
 @@@@
+   Attr.mk ~loc:Location.none local_ext_loc (PStr [])
+ 
+ let local_extension =
+   Exp.mk ~loc:Location.none (Pexp_extension(local_ext_loc, PStr []))
+ 
+-let include_functor_ext_loc = mknoloc "extension.include_functor"
```
Comment (goldfirere) [inline vendor/diff-parsers-ext-parsewyc.patch:186]: Why remove this include-functor stuff?

File: vendor/diff-parsers-ext-parsewyc.patch [hunk 10]
```diff
@@ -688,10 +807,41 @@
...
+       { Extensions.Comprehensions.Range { start = $2 ; stop = $4 ; direction = $3 } }
+@@@@
+ ;
+ 
+ %inline array_simple(ARR_OPEN, ARR_CLOSE, contents_semi_list):
+   | ARR_OPEN contents_semi_list ARR_CLOSE
+       { Generic_array.Literal $2 }
+-  | ARR_OPEN contents_semi_list error
+-      { Generic_array.Unclosed($loc($1),$loc($3)) }
++  (*| ARR_OPEN contents_semi_list error
```
Comment (goldfirere) [inline vendor/diff-parsers-ext-parsewyc.patch:822]: Why comment this out? Maybe this is all de rigueur for this file, but I'm pretty thrown off. If this is all normal for this stuff, you don't need to take the time to educate me here.

File: vendor/diff-parsers-upstream-std.patch [hunk 11]
```diff
@@ -133,8 +133,52 @@
      using an open recursion style: each method takes as its first
      argument the mapper to be applied to children in the syntax
      tree. *)
+--- ocaml-4.13-upstream/asttypes.mli
```
Comment (goldfirere) [inline vendor/diff-parsers-upstream-std.patch:136]: What's going on in this file? These changes seem unrelated to comprehensions and immutable arrays.

**Comment by ccasin** [comment]
Sorry I missed this when you posted it - busy week, I guess it slipped by in my email.  Looks like Richard has reviewed, but let me know if I can be helpful.

File: vendor/diff-parsers-ext-parsewyc.patch (hunk 9 above)
Comment (ccasin) [reply vendor/diff-parsers-ext-parsewyc.patch:186]: I think you should just ignore these patch files in your review.  They are generated files that just document differences from the upstream parser - no semantic effect.  

It's likely someone (me) forgot to commit them with a previous change.  This is fixed in the PR that cleans things up and rebases onto modern ocamlformat, but we haven't been able to merge that yet because of other formatting changes.  It should be fine to just merge whatever got generated here.

File: vendor/diff-parsers-ext-parsewyc.patch (hunk 8 above)
Comment (antalsz) [reply vendor/diff-parsers-ext-parsewyc.patch:689]: (See [Chris's comments about these patch files](https://github.com/janestreet/ocamlformat/pull/11#discussion_r1105980419) – resolving.)

File: vendor/diff-parsers-upstream-std.patch (hunk 11 above)
Comment (antalsz) [reply vendor/diff-parsers-upstream-std.patch:136]: (See [Chris's comments about these patch files](https://github.com/janestreet/ocamlformat/pull/11#discussion_r1105980419) – resolving.)

File: vendor/diff-parsers-ext-parsewyc.patch (hunk 10 above)
Comment (antalsz) [reply vendor/diff-parsers-ext-parsewyc.patch:822]: The file you actually wanted to review was `vendor/parse-wyc/lib/parser.mly` (see [Chris's comments about these patch files](https://github.com/janestreet/ocamlformat/pull/11#discussion_r1105980419)), not the patch file; indeed, if you look at the other parts of that file, you can see that the other error-handling pieces are also commented out.  I'm not sure why, but when in Rome, etc. etc.

File: lib/Ast.ml (hunk 2 above)
Comment (antalsz) [reply lib/Ast.ml]: Oh, that's a habit I adopted inconsistently for comments in `ocaml-jst` that were about the code differences; there, at least, the `JST` corresponds with the `-jst` in the name.  Less obvious here!  I think it's sort of intended to signify "here's some code that's just here to support being different", rather than "here's a change that could be upstreamed".  I think it's worth calling out, but that may not be the best prefix, yeah.

File: test/passing/tests/attributes.ml (hunk 4 above)
Comment (antalsz) [reply test/passing/tests/attributes.ml:85]: That's right, I'm just copying the surrounding code.  Resolving.

File: test/passing/tests/break_separators-after.ml.ref (hunk 5 above)
Comment (antalsz) [reply test/passing/tests/break_separators-after.ml.ref:94]: What you should be reviewing is how the immutable arrays look relative to the other arrays _in this file_.  If we look at the name of this file, it's `break_separators-after.ml.ref`; this means it comes from `break_separators.ml`, *subject to the settings denoted by `after`*.  So in *this* file, we're testing that the separators come "after" the list elements!  You'll likely want to take another look at this and the similar files with that in mind.

File: test/passing/tests/break_separators-before_docked_wrap.ml.ref (hunk 6 above)
Comment (antalsz) [reply test/passing/tests/break_separators-before_docked_wrap.ml.ref:105]: As with `test/passing/tests/break_separators-after.ml.ref`, but this time the "_wrap" part of the settings means that we're doing flow-based wrapping, I think.  Not resolving in case you want to look at other parts of this file.

File: test/passing/tests/break_separators-before_docked_wrap.ml.ref (hunk 7 above)
Comment (antalsz) [reply test/passing/tests/break_separators-before_docked_wrap.ml.ref:193]: See the other comment in this file; resolving.

File: vendor/diff-parsers-ext-parsewyc.patch (hunk 10 above)
Comment (ccasin) [reply vendor/diff-parsers-ext-parsewyc.patch:822]: I don't know the actual details, but the `parse-wyc` stuff is an experimental error recovery parser (parse-wyc is "parse what you can").  So it would make sense that it has different error handling code.

File: vendor/diff-parsers-ext-parsewyc.patch (hunk 9 above)
Comment (goldfirere) [reply vendor/diff-parsers-ext-parsewyc.patch:186]: OK.... but that's harder to review. That is, it sounds like you're suggesting that I examine e.g. parse-wyc/lib/parser.mly. Yet that file includes both the upstream changes and any local changes. The upstream changes, though, aren't really at issue here -- it's just the changes that are the diff between this repo and upstream that I would like to review. I can't easily spot these in the parser.mly file. However, I would imagine that this diff file, though autogenerated, would capture exactly the interesting bit.

It sounds, though, that this diff file in this PR captures aspects that aren't related to this PR, which also makes this file hard to review. So... :shrug:

File: lib/Ast.ml (hunk 3 above)
Comment (antalsz) [reply lib/Ast.ml:2617]: This doesn't seem any different than the usual `Extensions.Expression.of_ast` matches to me; what are you confused by?

Comment (goldfirere) [reply lib/Ast.ml:2617]: Not the `of_ast` specifically, but the `_ :: _` and `Yes, [e0]` bits. That is, it seems you're looking for some kind of source pattern to decide between `true` and `false` here for extensions, but the pattern you're looking for doesn't jump out at me.

Comment (antalsz) [reply lib/Ast.ml:2617]: Look right below, at lines 2616–2619 – it's the same pattern as the existing code.  (To wit: bracketed collections that have attributes (and are equal to `exp`).)

File: vendor/diff-parsers-ext-parsewyc.patch (hunk 9 above)
Comment (ccasin) [reply vendor/diff-parsers-ext-parsewyc.patch:186]: FWIW, I have typically found it easier to review the diff on the parser than the diff of diffs you are suggesting you want to review (because it typically requires the same context you need to review the parser itself).  But one way to enable the review you want would be to refactor this branch with a commit at the beginning that just includes the changes to the patch files that should have been there, and another commit at the end with all the other patch file changes (at some cost to Antal's sanity during the rebase).

File: lib/Ast.ml (hunk 3 above)
Comment (antalsz) [reply lib/Ast.ml:2617]: (I can update the comment, but it's `e0` because we can't compare with `ee0`, as that was synthesized.)

File: vendor/diff-parsers-ext-parsewyc.patch (hunk 9 above)
Comment (antalsz) [reply vendor/diff-parsers-ext-parsewyc.patch:186]: I’m perfectly happy to do that, that should be a free rebase

Comment (goldfirere) [reply vendor/diff-parsers-ext-parsewyc.patch:186]: I don't feel strongly at all here -- it's just that, as it currently sits, this content is unreviewed, mostly because I don't see a good way to review it. Either I look at this diff file, which contains a bunch of stuff unrelated to this patch; or I look at the change to the parser itself, which also contains a bunch of stuff unrelated to this patch (because it's in the upstream patch). If the actual content is tiny (or non-existent, because there are no actual changes related to this patch? possible.) and you're confident to let it go unreviewed, then that's fine by me.

File: lib/Ast.ml (hunk 3 above)
Comment (goldfirere) [reply lib/Ast.ml:2617]: OK. Yes. I see this now. A comment would be very helpful!

Comment (antalsz) [reply lib/Ast.ml:2617]: Added in b32eff2

File: vendor/diff-parsers-ext-parsewyc.patch (hunk 9 above)
Comment (antalsz) [reply vendor/diff-parsers-ext-parsewyc.patch:186]: The patch files are now available in 95c808b!

File: lib/Ast.ml (hunk 1 above)
Comment (antalsz) [reply lib/Ast.ml:1329]: 1. Adding an `| Eexp of Extensions.Expression.t` case?  That'll help in some places, but we still need all the regular machinery for when we don't have an `Ast.t`; I think that might end up confusing more than clarifying, since we'll get the invariant that the `Exp` case is the only expression we *don't* have to check.
2. `fun _ -> true` -- yes, this is wrong, it should be `fun _ -> false`.  (Is this pattern an expression child?  No, it is not.)  I could also make it a true type (`check_comprehension (Pattern pat)` vs. `check_comprehension (Expression exp)`); that would probably be clearest.  It can still be one function, though; the iteration structure is the same, I just have to get my identities right :-)
3. However, I contend the intended semantics of this function are correct: in `[x,y for x = 1 to 10 when x > 3 for y in create ()]`, the following are the direct children of the comprehension:
    * `x,y` (expression, body)
    * `x` (pattern, first iterator)
    * `1` (expression, first iterator)
    * `10` (expression, first iterator)
    * `x > 3` (expression, conditional)
    * `y` (pattern, second iterator)
    * `create ()` (expression, second iterator)
In particular, I think these have to be direct children because there's nothing else for them to be children of.
4. I suppose this could be what you mean by extending `Ast.t`: `| Citer of comprehension_iterator`.  This is probably doable, but I don't think it buys us anything; the fact that they're a child of a comprehension node is enough information to style them correctly, so it would be a lot of work for, as far as I can tell, no gain.

Comment (antalsz) [reply lib/Ast.ml:1329]: Please check out 569f51c and let me know if you're happier!

Comment (goldfirere) [reply lib/Ast.ml:1329]: Yes, very happy. Thanks! (And I fully agree about not doing the more-accurate child detection.)

File: vendor/diff-parsers-ext-parsewyc.patch (hunk 9 above)
Comment (goldfirere) [reply vendor/diff-parsers-ext-parsewyc.patch:186]: That commit looks good. Consider this reviewed. Thanks!

**Review by goldfirere — APPROVED** [review]

**Comment by goldfirere** [comment]
I don't want to make git unhappy by merging if you're not ready, but do merge if you're ready. Thanks!

**Comment by antalsz** [comment]
Ah, Carl requested that we not merge yet, I should've mentioned!  Next steps will be testing :-)  Thanks!
